src/tray/windows/SDL_tray.c

```diff
@@ -0,0 +1,589 @@
+/*
+  Simple DirectMedia Layer
+  Copyright (C) 1997-2024 Sam Lantinga <slouken@libsdl.org>
+
+  This software is provided 'as-is', without any express or implied
+  warranty.  In no event will the authors be held liable for any damages
+  arising from the use of this software.
+
+  Permission is granted to anyone to use this software for any purpose,
+  including commercial applications, and to alter it and redistribute it
+  freely, subject to the following restrictions:
+
+  1. The origin of this software must not be misrepresented; you must not
+     claim that you wrote the original software. If you use this software
+     in a product, an acknowledgment in the product documentation would be
+     appreciated but is not required.
+  2. Altered source versions must be plainly marked as such, and must not be
+     misrepresented as being the original software.
+  3. This notice may not be removed or altered from any source distribution.
+*/
+
+#include "SDL_internal.h"
+
+#include "../../video/windows/SDL_surface_utils.h"
+
+#include <windows.h>
+#include <windowsx.h>
+#include <shellapi.h>
+
+#include <stdlib.h>
+
+#define WM_TRAYICON (WM_USER + 1)
+
+struct SDL_TrayMenu {
+    HMENU hMenu;
+
+    size_t nEntries;
+    SDL_TrayEntry **entries;
+
+    SDL_Tray *parent_tray;
+    SDL_TrayEntry *parent_entry;
+};
+
+struct SDL_TrayEntry {
+    SDL_TrayMenu *parent;
+    UINT_PTR id;
+
+    char label_cache[4096];
+    SDL_TrayEntryFlags flags;
+    SDL_TrayCallback callback;
+    void *userdata;
+    SDL_TrayMenu *submenu;
+};
+
+struct SDL_Tray {
+    NOTIFYICONDATAW nid;
+    HWND hwnd;
+    HICON icon;
+    SDL_TrayMenu *menu;
+};
+
+static UINT_PTR get_next_id(void)
+{
+    static UINT_PTR next_id = 0;
+    return ++next_id;
+}
+
+static SDL_TrayEntry *find_entry_in_menu(SDL_TrayMenu *menu, UINT_PTR id)
+{
+    for (size_t i = 0; i < menu->nEntries; i++) {
+        SDL_TrayEntry *entry = menu->entries[i];
+
+        if (entry->id == id) {
+            return entry;
+        }
+
+        if (entry->submenu) {
+            SDL_TrayEntry *e = find_entry_in_menu(entry->submenu, id);
+
+            if (e) {
+                return e;
+            }
+        }
+    }
+
+    return NULL;
+}
+
+static SDL_TrayEntry *find_entry_with_id(SDL_Tray *tray, UINT_PTR id)
+{
+    if (!tray->menu) {
+        return NULL;
+    }
+
+    return find_entry_in_menu(tray->menu, id);
+}
+
+LRESULT CALLBACK TrayWindowProc(HWND hwnd, UINT uMsg, WPARAM wParam, LPARAM lParam) {
+    SDL_Tray *tray = (SDL_Tray *) GetWindowLongPtr(hwnd, GWLP_USERDATA);
+    SDL_TrayEntry *entry = NULL;
+
+    if (!tray) {
+        return DefWindowProc(hwnd, uMsg, wParam, lParam);
+    }
+
+    switch (uMsg) {
+        case WM_TRAYICON:
+            if (LOWORD(lParam) == WM_RBUTTONUP || LOWORD(lParam) == WM_LBUTTONUP) {
```

Comment:
For the custom callback, I'm not sure this would be easy for cross-platform support with non-Windows platforms.

I'll use `WM_CONTEXTMENU`, it seems important for proper keyboard support.

I'm not sure I understand correctly for `WM_USER` though. The [documentation](https://learn.microsoft.com/en-us/windows/win32/winmsg/wm-user) says that it marks the beginning of messages for use by the window class, and is not actually meaningful by itself. Do I have the correct `WM_USER` here?

Reply:
I've updated `RBUTTONUP` to `CONTEXTMENU`, but I'll check on my machine for the behaviour of the left-click before updating it. I'd like to search for some documentation about what's the best macro to use there.